## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [2]:
import pandas as pd
df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [3]:
# View the column names in the data
print(df.columns)


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [4]:
# Use the describe function to gather some basic statistics
statistics = df.describe()
print(statistics)


          client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.00000

In [5]:
# Use this space to do any additional research
# and familiarize yourself with the data.



In [6]:
# What three item categories had the most entries?
df["category"].value_counts().head(3)




category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [7]:
# For the category with the most entries,
# which subcategory had the most entries?

df.loc[df["category"] == "consumables", "subcategory"].value_counts().head(1)


subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [8]:
# Which five clients had the most entries in the data?

df['client_id'].value_counts().head(5)


client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [9]:
# Store the client ids of those top 5 clients in a list.
top_5_clients = list(df["client_id"].value_counts().head(5).index)
top_5_clients



[33615, 66037, 46820, 38378, 24741]

In [10]:
# How many total units (the qty column) did the
# client with the most entries order order?
df.loc[df["client_id"] == top_5_clients[0], "qty"].sum()


64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [11]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty

df["line_subtotal"] = df["unit_price"] * df["qty"]
df[['unit_price', 'qty', 'line_subtotal']].head(2)

unit_price  qty  line_subtotal
0     1096.80  105      115164.00
1       24.95   21         523.95

In [12]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.

def shipping_calculator(weight):
    if weight > 50:
        return weight * 7
    return weight * 10

df['total_weight'] = df['qty'] * df['unit_weight']
df['shipping_price'] = df['total_weight'].apply(shipping_calculator)
df[['unit_price', 'unit_weight', 'qty', 'total_weight', 'shipping_price']].head(3)

unit_price  unit_weight  qty  total_weight  shipping_price
0     1096.80         7.50  105        787.50         5512.50
1       24.95         1.49   21         31.29          312.90
2       13.52         1.68   39         65.52          458.64

In [13]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%
df["line_price"] = round((df["line_subtotal"] + df["shipping_price"]) * 1.0925, 2)
df[['line_subtotal', 'shipping_price', 'line_price']].head(3)  


line_subtotal  shipping_price  line_price
0      115164.00         5512.50   131839.08
1         523.95          312.90      914.26
2         527.28          458.64     1077.12

In [14]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).
df['line_cost'] = df['unit_cost'] * df['qty'] + df['shipping_price']
df.head(3)


first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   

   line_number  line_subtotal  total_weight  shipping_price  line_price  \
0            1      115164.00        787.50         5512.50   131839.08   
1            0         523.95         31.29          312.90      914.26   
2            6         527.28         65.52          458.64     1077.12   

   line_cost  
0   85597.05  
1     629.79  
2     765.18  

[3 rows x 23 columns]

In [15]:
# Create a column for the profit of
# each line using line cost and line price
df["line_profit"] = df["line_price"] - df["line_cost"]
df.head(3)




first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   

   line_subtotal  total_weight  shipping_price  line_price  line_cost  \
0      115164.00        787.50         5512.50   131839.08   85597.05   
1         523.95         31.29          312.90      914.26     629.79   
2         527.28         65.52          458.64     1077.12     765.18   

   line_profit  
0     46242.03  
1       284.47  
2       311.94  

[3 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [16]:
# Check your work using the totals above
import numpy as np

order_ids = [2742071, 2173913, 6128929]

def find_order_price(order_id):
    order_lines_df = df.loc[df['order_id'] == order_id, 'line_price']
    return round(order_lines_df.sum(), 2)

for order_id in order_ids:
    print(f"Order {order_id} total: ${find_order_price(order_id)}")

Order 2742071 total: $152811.89
Order 2173913 total: $162388.71
Order 6128929 total: $923441.25


In [17]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.

def find_client_id(client_id,column):
    client_df = df.loc[df['client_id'] == client_id, column]
    return round(client_df.sum(), 2)

for client_id in top_5_clients:
    print(f"{client_id} : ${find_client_id(client_id, 'line_price')}")
    









33615 : $8377308.52
66037 : $10259514.79
46820 : $9743794.36
38378 : $12906550.87
24741 : $82268892.02


In [18]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.

# Select the top 5 clients
top_5_clients_df = df[df["client_id"].isin(top_5_clients)]

# Group by client id
top_5_clients_grouped = top_5_clients_df.groupby("client_id")
top_5_clients_grouped = top_5_clients_grouped.agg({ "qty": "sum", "shipping_price": "sum", "line_price": "sum", "line_profit": "sum" })
top_5_clients_grouped = top_5_clients_grouped.sort_values("line_profit", ascending=False)   
top_5_clients_grouped.head(5)






qty  shipping_price   line_price  line_profit
client_id                                                  
24741      239862      5126448.37  82268892.02  36579992.31
38378       73667      3429455.40  12906550.87   3271829.89
66037       43018      1395151.85  10259514.79   3255031.81
46820       75768      1601448.84   9743794.36   2736602.72
33615       64313      1828984.89   8377308.52   2201994.61

In [19]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.

df = df.rename(columns={"qty": "Total Units Purchased", "shipping_price": "Total Shipping Price", "line_price": "Total Revenue", "line_profit": "Total Profit"})
df["Total Shipping Price"] = df["Total Shipping Price"].map("${:,.2f}".format)  
df["Total Revenue"] = df["Total Revenue"].map("${:,.2f}".format)
df["Total Profit"] = df["Total Profit"].map("${:,.2f}".format)
df["Total Units Purchased"] = df["Total Units Purchased"].map("{:,}".format)
df.head(5)





first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight Total Units Purchased  \
0          17        2023  ...    762.71        7.50                   105   
1          20        2023  ...     15.09        1.49                    21   
2           4        2023  ...      7.86        1.68                    39   
3          17        2023  ...     24.85        1.23                    29   
4          21        2023  ...    108.17       46.43                    20   

   line_number  line_subtotal  total_weight Total Shipping Price  \
0            1      115164.00        787.50            $5,512.50   
1            0         523.95         31.29              $312.90   
2            6         527.28         65.52              $458.64   
3            3        1056.18         35.67              $356.70   
4            1        3902.00        928.60            $6,500.20   

   Total Revenue  line_cost  Total Profit  
0    $131,839.08   85597.05    $46,242.03  
1        $914.26     629.79       $284.47  
2      $1,077.12     765.18       $311.94  
3      $1,543.57    1077.35       $466.22  
4     $11,364.40    8663.60     $2,700.80  

[5 rows x 24 columns]

In [27]:
# Sort the updated data by "Total Profit" form highest to lowest
df = df.sort_values("Total Profit", ascending=False)
df.head(30)

first       last                                          job  \
18985      Charles  Mccormick     Conservation officer, historic buildings   
22487        Jesus     Wright                           Theme park manager   
2990   Christopher      Ortiz                                   Printmaker   
34960      Jessica   Gonzalez                          Editorial assistant   
45220         Mark   Williams                             Fashion designer   
23492         Paul    Vasquez           Teacher, special educational needs   
7711          Jose   Martinez                      Physiological scientist   
15887      Heather        Day                   Regulatory affairs officer   
34188          Amy      Leach                      Data processing manager   
21636    Alexandra      Young                                    Dietitian   
43305        Jared     Taylor                                Haematologist   
42596      Patrick    Spencer                    Insurance account manager   
50485        James  Wilkinson                  Conservator, museum/gallery   
54225       Nathan    Camacho                          Scientist, forensic   
5964       Richard    Kennedy                       Race relations officer   
17357         Sean     Steele                          Administrator, arts   
42163        Scott     Harper                           Personal assistant   
720          James     Atkins                                 Sports coach   
14914       Donald    Harding                                 Immunologist   
43752          Ian        Fox                   Lecturer, higher education   
47779      Michael     Torres                    Medical technical officer   
25914       Thomas      Ellis                                Haematologist   
46964       Ashley       Hill                          Automotive engineer   
40134      Caitlyn     Golden                        Private music teacher   
40378        Barry    Johnson                    Education officer, museum   
24364       Nicole     Martin  Armed forces training and education officer   
18870       Briana       Luna                                    Press sub   
27994        Emily      White                              Producer, radio   
41815       Elaine   Valencia                        Glass blower/designer   
54228       Donald    Simpson                          Information officer   

                        phone                                   email  \
18985   001-667-790-6949x2885       mccormick_sign@potter-sanford.com   
22487      605.947.9467x66513           wright.jesus.7579@meadows.com   
2990        759.426.7676x6963                 c.ortiz.9576@morris.com   
34960  001-766-709-4338x01962               gonzalezj@horn-fields.com   
45220           (589)855-5294          home.williams@reed-sanchez.com   
23492      (520)767-2543x5513        paul_vasquez@brown-middleton.biz   
7711      +1-936-839-4988x476           martinezjose6115@aguilar.info   
15887       (251)730-3328x301                         dayh@hughes.net   
34188     (429)706-0482x48605         leach.bring.5704@myers-snow.com   
21636        265-666-3155x118                  ayoung9158@stevens.com   
43305            729-249-1032                      taylorj@romero.com   
42596        407.863.3652x665        spencersupport@rogers-harris.biz   
50485        991.297.5872x313        jamesperson@gonzalez-hoffman.biz   
54225    001-274-275-1688x481                 nathancamacho@brown.com   
5964          +1-551-603-4716                    rkennedy@mcneil.info   
17357      698.307.8917x62247        seansteele@coleman-schroeder.com   
42163              4583749266                  areas3327@copeland.net   
720              677.418.8364              atkins.huge.354@deleon.com   
14914      793-904-7725x39308        harding.donald.7185@sullivan.com   
43752           (917)710-6686                    i.fox.6520@terry.biz   
47779        001-772-485-0471   michael.torres.9257@oneal-dou